vispy/visuals/visual.py

```diff
@@ -347,34 +349,109 @@ def set_gl_state(self, preset=None, **kwargs):
 
         The arguments are forwarded to :func:`vispy.gloo.wrappers.set_state`.
 
+        This can also be used as a context manager that will revert the
+        gl_state on exit.
+
         Parameters
         ----------
         preset : str
             Preset to use.
         **kwargs : dict
             Keyword arguments.
         """
+        prev_gl_state = self._vshare.gl_state.copy()
+
         self._vshare.gl_state = kwargs
         self._vshare.gl_state['preset'] = preset
```

Comment:
I'm sure this is unlikely as far as a common use case, but I think technically this state modification has to happen in the `_context` function before the `yield`. Otherwise this would not work:

```python
state_context = vis.set_gl_state(...)
# do some stuff with old state
...
with state_context:
    # do some stuff with new state
    ...
```

Reply:
Hm, that's a good point but I need to play with this a bit more then. I don't think it gets called in that case if you never enter the context. I wrote it this way to try to preserve the current behavior as much as possible but still use the existing function names.

Reply:
I guess it could happen in both places? It's redundant work in the case of using a context manager but not too costly.

Reply:
Oh I forgot about the old behavior. You're absolutely right. Leave it like you have it then.